non-custodial payouts extension (#203)
stratum-mining/sv2-spec: extensions/0x0003-non-custodial-payouts.md

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119).

---

## 1. Motivation

Base JDP cannot support non-custodial payouts due to an ordering constraint:

1. `AllocateMiningJobToken.Success` fixes the coinbase outputs (and thus the space/sigops JDC reserves via `CoinbaseOutputConstraints`).
2. Only then can JDC receive `NewTemplate`, which is when the template revenue `T` becomes known.

A multi-miner payout requires amounts that depend on `T`, so the outputs cannot be fixed in step 1. This extension resolves the paradox by having JDS publish a **weight-based payout distribution**: outputs and relative weights are known upfront (fixing space/sigops), while absolute amounts are computed deterministically once `T` is known.

---

## 2. Negotiation

The extension is negotiated via [`RequestExtensions`](./0x0001-extensions-negotiation.md) (extension `0x0001`) with extension identifier `0x0003`.

> JDC MUST negotiate this extension on **both** the JDP connection with JDS and the Mining Protocol connection with Pool. If negotiation succeeds on only one connection, JDC MUST NOT use this extension, and Pool/JDS MUST reject any job declaration carrying this extension's TLV fields from such a client.

Reviewer: ```suggestion
JDC MUST negotiate this extension on **both** the Job Declaration Protocol (JDP) connection with JDS and the Mining Protocol (MP) connection with Pool. If negotiation succeeds on only one connection, JDC MUST NOT use this extension, and Pool/JDS MUST reject any job declaration carrying this extension's TLV fields from such a client.
```

Author: done

| Field Name          | Data Type        | Description                                                                                                |
| ------------------- | ---------------- | ---------------------------------------------------------------------------------------------------------- |
| distribution_id     | U64              | Strictly increasing identifier, universal across all Pool/JDS connections of this Pool.                   |
| pool_payout         | B0_64K           | Consensus-serialized tx output. Locking script MUST be controlled by Pool. Amount field MUST be non-0 and encodes `weight_P`, a relative weight (not satoshis). |
| payouts             | SEQ0_64K[B0_64K] | Consensus-serialized tx outputs for miners. Amount fields MUST be non-0 and encode relative weights `weight[i]` (not satoshis). |
| dust_limits         | SEQ0_64K[U32]    | `dust_limits[i]` (satoshis) for each `payouts[i]`. MUST have the same length as `payouts`.                        |

> | additional_outputs  | SEQ0_64K[B0_64K] | Consensus-serialized tx outputs added by Pool (e.g. `OP_RETURN`). Amount fields MUST be 0.                 |

Reviewer: As currently written, a reader will assume `weight_P` is just the pool fee. It should be explicit that it also carries value credited for miners that don't get their own output.

```suggestion
| additional_outputs  | SEQ0_64K[B0_64K] | Consensus-serialized tx outputs added by Pool (e.g. `OP_RETURN`). Amount fields MUST be 0.                 |

`weight_P` is not necessarily just the Pool's fee. Besides the Pool's own share, it MAY include value credited on behalf of miners that are not represented in `payouts` — for example because their pending earnings are below the Pool's payout threshold, because the Pool limits the number of payout outputs in the coinbase, or because of any similar pool policy. Value earned by such miners is paid into the Pool's output; how the Pool tracks it, and how those miners are eventually included in a future distribution, is pool policy and out of scope for this extension.
```

Author: done

#### 3.1.1 Update Policy

Share-based accounting methods shift weights with every submitted share. JDS MUST NOT publish a new `SetPayoutDistribution` per share; it SHOULD only publish periodically (e.g. on a timer, or when weights change materially).

---

## 4. Payout Computation

> Upon receiving a `NewTemplate`, JDC builds the coinbase tx outputs based on the latest `SetPayoutDistribution` it received.

Reviewer: "based on the latest" is currently descriptive, not normative. Making it a MUST turns deliberate distribution shopping ([§9.1]) into a protocol violation instead of tolerated behavior. It needs one carve-out for the §5 constraint transition.

```suggestion
Upon receiving a `NewTemplate`, JDC builds the coinbase tx outputs based on the latest `SetPayoutDistribution` it received.

Upon receiving a new `SetPayoutDistribution`, JDC MUST use it as the basis for all subsequently declared jobs. The only exception is the constraint transition of [§5](#5-coinbaseoutputconstraints): while waiting for a `NewTemplate` honoring updated constraints, JDC keeps declaring under the previous distribution.
```

Author: incorporating a slightly different form:
```diff
--- a/extensions/0x0003-non-custodial-payouts.md
+++ b/extensions/0x0003-non-custodial-payouts.md
@@ -60,7 +60,9 @@ Share-based accounting methods shift weights with every submitted share. JDS MUS

 ## 4. Payout Computation

-Upon receiving a `NewTemplate`, JDC builds the coinbase tx outputs based on the latest `SetPayoutDistribution` it received.
+Upon receiving a `NewTemplate`, JDC builds the coinbase tx outputs based on the latest `SetPayoutDistribution` it received, subject to the update rule below.
+
+Upon receiving a new `SetPayoutDistribution`, JDC MUST use it as the basis for all subsequently declared jobs. The only exception is the constraint transition of [§5](#5-coinbaseoutputconstraints): while waiting for a `NewTemplate` honoring updated constraints, JDC keeps declaring under the previous distribution.

 Variables:

@@ -180,7 +182,7 @@ Responses carrying `stale-payout-distribution` SHOULD NOT trigger the fallback s

 ### 9.1 Distribution Shopping

-If JDS implements the grace window ([§7.2](#72-grace-window)), a JDC holding distributions `k-1` and `k` can choose whichever pays it more. This is bounded to one window: the Pool settles per the `distribution_id` of the winning declaration, and the periodic update policy ([§3.1.1](#311-update-policy)) keeps the option value small.
+If JDS implements the grace window ([§7.2](#72-grace-window)), a JDC holding distributions `k-1` and `k` can choose whichever pays it more (a violation of the [§4](#4-payout-computation) update rule, but indistinguishable from honest races on the wire). This is bounded to one window: the Pool settles per the `distribution_id` of the winning declaration, and the periodic update policy ([§3.1.1](#311-update-policy)) keeps the option value small.

 ### 9.2 Privacy
```

| `weight[i]`      | `SetPayoutDistribution.payouts`           | relative weight for miner `i`                      |
| `weight_P`       | `SetPayoutDistribution.pool_payout`       | relative weight for pool-controlled address        |
| `W`              | `weight_P + Σ weight[i]`                  | sum total of all weights                           |
| `amount[i]`      | computed                                  | miner `i` proportional amount, before dust pruning |
| `amount_P`       | computed                                  | pool proportional amount                           |
| `pay[i]`         | computed                                  | final miner `i` payout (satoshis); no output if `amount[i] < dust_limits[i]` (dust-pruned) |
| `pay_P`          | computed                                  | final Pool payout (satoshis); absorbs integer rounding and dust pruning   |

Formulae:

```
W         = weight_P + Σ weight[i]
amount_P  = floor(weight_P · T / W)
amount[i] = floor(weight[i] · T / W)
pay[i]    = if amount[i] ≥ dust_limits[i] { amount[i] } else { dust-pruned }
pay_P     = T − Σ pay[i]
```

> There's no `dust_limit` corresponding to `pool_payout`, therefore `pay_P` MUST NOT be dust-pruned.

Reviewer: Worth spelling out the full composition of `pay_P` here, since this is where the reader is doing the math. Only the dust/rounding component is currently mentioned (in the §4 variable table).

```suggestion
There's no `dust_limit` corresponding to `pool_payout`, therefore `pay_P` MUST NOT be dust-pruned.

`pay_P` therefore aggregates three components:

1. the Pool's own share, via `weight_P`;
2. value credited on behalf of miners not represented in `payouts` (see [§3.1](#31-setpayoutdistribution-jds---jdc)), also via `weight_P`;
3. dust-pruned amounts and integer rounding remainders, absorbed at payout-computation time.

Components 2 and 3 are miner-earned value that could not be paid on-chain in this block.
```

Author: done

When this extension is negotiated, JDC MUST append this TLV field to the following message, depending on the Job Declaration mode:

| Mode           | Message                                 |
| -------------- | --------------------------------------- |
| Full-Template  | `DeclareMiningJob`                      |
| Coinbase-only  | `SetCustomMiningJob`                    |

---

## 7. Validation

### 7.1 Output Verification

Validation is recompute-and-compare: the verifier recomputes the expected output vector from (`SetPayoutDistribution`, `T`) per [§4](#4-payout-computation) and compares it against the declared coinbase, allowing only JDC-appended 0-value outputs in position 4.

### 7.2 Grace Window

JDS/Pool MAY accept declarations matching the distribution immediately before the current one (a grace window). This absorbs honest races — declarations already in flight when a new distribution is published — at the cost of one-window optionality for miners ([§9.1](#91-distribution-shopping)). Any declaration not covered by the acceptance window MUST be rejected with the `stale-payout-distribution` error code.

> **Settlement exception**: when the Pool finds a block, it settles its accounting window based on the `distribution_id` of the declaration that produced the winning block. At settlement, JDS atomically invalidates all prior distributions and publishes a fresh `SetPayoutDistribution` (`distribution_id = k`); only `k` is accepted until `k+1` is published, after which the grace window (if any) resumes. In-flight declarations on invalidated distributions are rejected with `stale-payout-distribution`. This prevents a block found on a pre-settlement distribution from paying an already-settled window twice.

Reviewer: Do you think this is really needed? Isn't this up to the implementation?

Author: hmm that's a tricky question to answer

IMO the sentence above (which establishes the notion of the grace window) is important, beacuse it establishes an "official" mechanism that mitigates honest races between `SetPayoutDistribution` (JDS->JDC) and `NewTemplate` (TP->JDC), in a way that fosters interoperability across different implementations

now about the sentence starting at line 146: 

its main purpose is more like a warning against footguns, than something that spec is trying to impose.

we don't want pools that opt-into the grace window to settle blindly with regards to which `distribution_id` is associated with the block that was found, because that could mess up their accounting.

but I agree it's overstating things a bit, so I'm moving it to section `10. Implementation Notes` and using `SHOULD` language

Reviewer: I analyzed this with Fable, and I left some comments which are related to these points.

This extension defines the following `error_code` values for `DeclareMiningJob.Error` and `SetCustomMiningJob.Error`:

| error_code                  | Description                                                              |
| --------------------------- | ------------------------------------------------------------------------ |
| `stale-payout-distribution`        | `distribution_id` not accepted (too old, or invalidated by settlement — see [§10](#10-implementation-notes)). |
| `invalid-payout-distribution` | Coinbase outputs violate [§4](#4-payout-computation).                                  |

> Responses carrying `stale-payout-distribution` SHOULD NOT trigger the fallback strategy of [JDP §6.2](../06-Job-Declaration-Protocol.md#62-job-declarator-client) — they cover honest races (in-flight declarations).

Reviewer: ```suggestion
Responses carrying `stale-payout-distribution` SHOULD NOT trigger the fallback strategy of [JDP §6.2](../06-Job-Declaration-Protocol.md#62-job-declarator-client) — they cover honest races (in-flight declarations). This exemption is not unconditional: an honest race resolves once JDC re-declares against the latest received distribution, so JDC SHOULD treat persistent `stale-payout-distribution` rejections of declarations built on the latest received distribution as it would any other rejection, and apply the fallback strategy of [JDP §6.2](../06-Job-Declaration-Protocol.md#62-job-declarator-client).
```

Author: done

---

## 8. Message Types

| Message Type (8-bit) | channel_msg_bit | Message Name           |
| -------------------- | --------------- | ---------------------- |
| 0x00                 | 0               | SetPayoutDistribution  |

**Message Framing:** `SetPayoutDistribution` MUST have `extension_type = 0x0003` in its frame header. `DeclareMiningJob` and `SetCustomMiningJob` keep `extension_type = 0x0000` (modified via TLV only). See [§3.4.1 of the Protocol Overview](../03-Protocol-Overview.md#341-extension-type-field-usage).

---

## 9. Security Considerations

### 9.1 Distribution Shopping

If JDS implements the grace window ([§7.2](#72-grace-window)), a JDC holding distributions `k-1` and `k` can choose whichever pays it more. This is bounded to one window: the Pool settles per the `distribution_id` of the winning declaration, and the periodic update policy ([§3.1.1](#311-update-policy)) keeps the option value small.

### 9.2 Privacy

The distribution is pool-wide: every JDC learns the payout scripts and relative weights of the Pool's top miners. Pools SHOULD weigh this leak when choosing the number of payout slots.

### 9.3 Trust Model

> The Pool still dictates scripts and weights, so custodial trust assumptions are unchanged. However, miners included in the distribution gain coinbase-level verifiability of their payout, and dust-pruned/rounded amounts are auditable (they flow to `pay_P` by construction).

Reviewer: "Auditable" tells the reader where the money went, but not what it *is*. The trust model should state plainly that value which can't be paid on-chain (pre-credited into `weight_P`, or absorbed by `pay_P` at computation time) sits under custodial trust assumptions — its later crediting is pool policy, outside this extension's scope.

```suggestion
The Pool still dictates scripts and weights, so custodial trust assumptions are unchanged. However, miners included in the distribution gain coinbase-level verifiability of their payout, and dust-pruned/rounded amounts are auditable (they flow to `pay_P` by construction).

Value that cannot be paid on-chain — whether credited into `weight_P` when the distribution is built (miners below the Pool's payout threshold, output-count caps, or similar policies) or absorbed by `pay_P` at payout-computation time (dust pruning, integer rounding) — is held by the Pool. For this value, the trust assumptions are the same as under custodial payouts: whether and how it is credited back to the affected miners is pool accounting policy, outside the scope of this extension. As a mitigation, the composition of `pay_P` is fully computable by any JDC holding the distribution, so the held amounts are auditable to the satoshi.
```

Author: done

---

## 10. Implementation Notes

Pool and JDS are assumed to share state.

> **Settlement invalidation**: when the Pool finds a block, it settles its accounting per the `distribution_id` of the winning declaration. Pool/JDS SHOULD atomically invalidate all other distributions and publish a fresh `SetPayoutDistribution`; Pool/JDS SHOULD only accept job declarations based on this new distribution, until the next one is published. This prevents a block found on a pre-settlement distribution from paying an already-settled window twice. Declarations rejected for this reason carry `stale-payout-distribution` and are honest races ([§7.3](#73-error-codes)).

Reviewer: ```suggestion
**Settlement invalidation**: when the Pool finds a block, it settles its accounting per the `distribution_id` of the winning declaration. Pool/JDS SHOULD atomically invalidate all other distributions and publish a fresh `SetPayoutDistribution`; Pool/JDS SHOULD only accept job declarations based on this new distribution, until the next one is published (the grace window of [§7.2](#72-grace-window) SHOULD NOT span a settlement event). This prevents *new* declarations from paying an already-settled window twice. It cannot, however, prevent a block being found on a job already declared and accepted before settlement: such a block pays the settled distribution again, and since payouts are non-custodial, this cannot be corrected after the fact. How this risk is absorbed is pool accounting policy, out of scope for this extension. Declarations rejected due to settlement invalidation carry `stale-payout-distribution` and are honest races ([§7.3](#73-error-codes)).
```

Author: done

---

## 8. Message Types

| Message Type (8-bit) | channel_msg_bit | Message Name           |
| -------------------- | --------------- | ---------------------- |
| 0x00                 | 0               | SetPayoutDistribution  |

**Message Framing:** `SetPayoutDistribution` MUST have `extension_type = 0x0003` in its frame header. `DeclareMiningJob` and `SetCustomMiningJob` keep `extension_type = 0x0000` (modified via TLV only). See [§3.4.1 of the Protocol Overview](../03-Protocol-Overview.md#341-extension-type-field-usage).

---

## 9. Security Considerations

### 9.1 Distribution Shopping

> If JDS implements the grace window ([§7.2](#72-grace-window)), a JDC holding distributions `k-1` and `k` can choose whichever pays it more (a violation of the [§4](#4-payout-computation) update rule, but indistinguishable from honest races on the wire). This is bounded to one window: the Pool settles per the `distribution_id` of the winning declaration, and the periodic update policy ([§3.1.1](#311-update-policy)) keeps the option value small.

Reviewer: ```suggestion
If JDS implements the grace window ([§7.2](#72-grace-window)), a JDC holding distributions `k-1` and `k` can choose whichever pays it more (a violation of the [§4](#4-payout-computation) update rule, but indistinguishable from honest races on the wire). This is bounded to one window: the Pool settles per the `distribution_id` of the winning declaration, and a sufficiently short update interval ([§3.1.1](#311-update-policy)) keeps the option value small.
```

Author: this is a good improvement, but it's also correlated with section `3.1.1`

introducing this change:
```diff
diff --git a/extensions/0x0003-non-custodial-payouts.md b/extensions/0x0003-non-custodial-payouts.md
index fa0155e..aae4401 100644
--- a/extensions/0x0003-non-custodial-payouts.md
+++ b/extensions/0x0003-non-custodial-payouts.md
@@ -56,6 +56,8 @@ MUST be the first message JDS sends after `SetupConnection.Success` and `Request

 Share-based accounting methods shift weights with every submitted share. JDS MUST NOT publish a new `SetPayoutDistribution` per share; it SHOULD only publish periodically (e.g. on a timer, or when weights change materially).

+The interval length is a pool-tunable tradeoff: shorter intervals reduce the distribution-shopping option value under the optional grace window ([§9.1](#91-distribution-shopping)); longer intervals reduce job rollover overhead.
+
 ---

 ## 4. Payout Computation
@@ -182,7 +184,7 @@ Responses carrying `stale-payout-distribution` SHOULD NOT trigger the fallback s

 ### 9.1 Distribution Shopping

-If JDS implements the grace window ([§7.2](#72-grace-window)), a JDC holding distributions `k-1` and `k` can choose whichever pays it more (a violation of the [§4](#4-payout-computation) update rule, but indistinguishable from honest races on the wire). This is bounded to one window: the Pool settles per the `distribution_id` of the winning declaration, and the periodic update policy ([§3.1.1](#311-update-policy)) keeps the option value small.
+If JDS implements the grace window ([§7.2](#72-grace-window)), a JDC holding distributions `k-1` and `k` can choose whichever pays it more (a violation of the [§4](#4-payout-computation) update rule, but indistinguishable from honest races on the wire). This is bounded to one window: the Pool settles per the `distribution_id` of the winning declaration, and a sufficiently short update interval ([§3.1.1](#311-update-policy)) keeps the option value small.

 ### 9.2 Privacy

```

Reviewer: perfect, thanks

> # Stratum V2 Extension: Non-Custodial Payouts

Reviewer 2: I would not call this non-custodial given it still requires full trust in the pool to not give you bad payout info. Maybe just "Coinbase Transaction Payouts"?
